src/sql/workbench/services/objectExplorer/browser/treeUpdateUtils.ts

```diff
@@ -281,14 +294,31 @@ export class TreeUpdateUtils {
 					showFirewallRuleOnError: true,
 					showDashboard: false
 				};
+				const expansionTimeoutValueSec = configurationService.getValue<number>(NODE_EXPANSION_CONFIG);
 				// Need to wait for the OE service to update its nodes in order to resolve the children
 				const nodesUpdatedPromise = new Promise((resolve, reject) => {
-					objectExplorerService.onUpdateObjectExplorerNodes(e => {
-						if (e.errorMessage) {
-							reject(new Error(e.errorMessage));
-						}
-						if (e.connection.id === connection.id) {
-							resolve(undefined);
+					// Clean up timeout and listener
+					const cleanup = () => {
+						clearTimeout(nodeUpdateTimer);
+						nodesUpdatedListener.dispose();
+					}
+
+					// If the node update takes too long, reject the promise
+					const nodeUpdateTimeout = () => {
+						reject(new Error(nls.localize('objectExplorerTimeout', "Object Explorer timed out for {0}", connection.databaseName)));
+						cleanup();
+					}
+					const nodeUpdateTimer = setTimeout(nodeUpdateTimeout, expansionTimeoutValueSec * 1000);
+
+
+					const nodesUpdatedListener = objectExplorerService.onUpdateObjectExplorerNodes(e => {
+						if (e.connection && e.connection.id === connection.id) {
+							if (e.errorMessage) {
+								reject(new Error(e.errorMessage));
+							} else {
+								resolve(undefined);
```

Comment:
If you specify this as a `Promise<void>` then you can get rid of the undefined here - since you're not using the return value anyways.